platforms/esp/32/clockless_i2s_esp32.h

```diff
@@ -0,0 +1,767 @@
+static int CLOCK_DIVIDER_A;
+static int CLOCK_DIVIDER_B;
+
+template <int DATA_PIN, int T1, int T2, int T3, EOrder RGB_ORDER = RGB, int XTRA0 = 0, bool FLIP = false, int WAIT_TIME = 5>
+class ClocklessController : public CPixelLEDController<RGB_ORDER>
+{
+    // -- Store the GPIO pin
+    gpio_num_t     mPin;
+    
+    // -- This instantiation forces a check on the pin choice
+    FastPin<DATA_PIN> mFastPin;
+    
+    // -- Save the pixel controller
+    PixelController<RGB_ORDER> * mPixels;
+    
+public:
+
+    void init()
+    {
+        i2sInit();
+        
+        // -- Allocate space to save the pixel controller
+        //    during parallel output
+        mPixels = (PixelController<RGB_ORDER> *) malloc(sizeof(PixelController<RGB_ORDER>));
+        
+        gControllers[gNumControllers] = this;
+        int my_index = gNumControllers;
+        gNumControllers++;
+        
+        // -- Set up the pin We have to do two things: configure the
+        //    actual GPIO pin, and route the output from the default
+        //    pin (determined by the I2S device) to the pin we
+        //    want. We compute the default pin using the index of this
+        //    controller in the array. This order is crucial because
+        //    the bits must go into the DMA buffer in the same order.
+        mPin = gpio_num_t(DATA_PIN);
+        
+        PIN_FUNC_SELECT(GPIO_PIN_MUX_REG[DATA_PIN], PIN_FUNC_GPIO);
+        gpio_set_direction(mPin, (gpio_mode_t)GPIO_MODE_DEF_OUTPUT);
+        pinMode(mPin,OUTPUT);
+        gpio_matrix_out(mPin, i2s_base_pin_index + my_index, false, false);
+    }
+    
+    virtual uint16_t getMaxRefreshRate() const { return 400; }
+    
+protected:
+   
+   static int pgcd(int smallest,int precision,int a,int b,int c)
+    {
+        int pgc_=1;
+        for( int i=smallest;i>0;i--)
+        {
+            
+            if( a%i<=precision && b%i<=precision && c%i<=precision)
+            {
+                pgc_=i;
+                break;
+            }
+        }
+        return pgc_;
+    }
+    
+    /** Compute pules/bit patterns
+     *
+     *  This is Yves Bazin's mad code for computing the pulse pattern
+     *  and clock timing given the target signal given by T1, T2, and
+     *  T3. In general, these parameters are interpreted as follows:
+     *
+     *  a "1" bit is encoded by setting the pin HIGH to T1+T2 ns, then LOW for T3 ns
+     *  a "0" bit is encoded by setting the pin HIGH to T1 ns, then LOW for T2+T3 ns
+     *
+     */
+    static void initBitPatterns()
+    {
+        // Precompute the bit patterns based on the I2S sample rate
+        Serial.println("Setting up fastled using I2S");
+
+        // -- First, convert back to ns from CPU clocks
+        uint32_t T1ns = ESPCLKS_TO_NS(T1);
+        uint32_t T2ns = ESPCLKS_TO_NS(T2);
+        uint32_t T3ns = ESPCLKS_TO_NS(T3);
+        
+        Serial.print("T1 = "); Serial.print(T1); Serial.print(" ns "); Serial.println(T1ns);
```

Comment:
Serial.print's should either be commented out or gated by #ifdef or a variable (I try to not have the library generally writing to the Serial port - for cases where folks are using the Serial port as, say, a control channel).